src/Swift.Bindings/tests/CryptoKit/Foundation.Source.cs

```diff
@@ -0,0 +1,274 @@
+
+        [DllImport(Path, EntryPoint = "$s10Foundation4DataV5countSivg")]
+        [UnmanagedCallConv(CallConvs = [ typeof(CallConvSwift) ])]
+        public static unsafe extern nint PInvoke_Data_GetCount(Data data);
+
+        [DllImport(Path, EntryPoint = "$s10Foundation4DataV9copyBytes2to5countySpys5UInt8VG_SitF")]
+        [UnmanagedCallConv(CallConvs = [ typeof(CallConvSwift) ])]
+        public static unsafe extern void PInvoke_Data_CopyBytes(UnsafeMutablePointer<byte> buffer, nint count, Data data);
+
+        [DllImport(Path, EntryPoint = "swift_getWitnessTable")]
+        [UnmanagedCallConv(CallConvs = [ typeof(CallConvSwift) ])]
+        public static unsafe extern void* PInvoke_Swift_GetWitnessTable(void* conformanceDescriptor, void* typeMetadata, void* instantiationArgs);
+
+        [DllImport(Path, EntryPoint = "$s10Foundation4DataVMa")]
+        [UnmanagedCallConv(CallConvs = [ typeof(CallConvSwift) ])]
+        public static unsafe extern void* PInvoke_Data_GetMetadata();
+    }
+
+    /// <summary>
+    /// Swift runtime helper methods in C#.
+    /// </summary>
+    public static class Runtime
+    {
+        /// <summary>
+        /// https://github.com/apple/swift/blob/main/include/swift/ABI/MetadataValues.h#L117
+        /// </summary>
+        [Flags]
+        public enum ValueWitnessFlags
+        {
+            AlignmentMask = 0x0000FFFF,
+            IsNonPOD = 0x00010000,
+            IsNonInline = 0x00020000,
+            HasSpareBits = 0x00080000,
+            IsNonBitwiseTakable = 0x00100000,
+            HasEnumWitnesses = 0x00200000,
+            Incomplete = 0x00400000,
+        }
+
+        /// <summary>
+        /// See https://github.com/apple/swift/blob/main/include/swift/ABI/ValueWitness.def
+        /// </summary>
+        [StructLayout (LayoutKind.Sequential)]
+        public ref struct ValueWitnessTable
+        {
+            public IntPtr InitializeBufferWithCopyOfBuffer;
+            public IntPtr Destroy;
+            public IntPtr InitWithCopy;
+            public IntPtr AssignWithCopy;
+            public IntPtr InitWithTake;
+            public IntPtr AssignWithTake;
+            public IntPtr GetEnumTagSinglePayload;
+            public IntPtr StoreEnumTagSinglePayload;
+            private IntPtr _Size;
+            private IntPtr _Stride;
+            public ValueWitnessFlags Flags;
+            public uint ExtraInhabitantCount;
+            public int Size => _Size.ToInt32();
+            public int Stride => _Stride.ToInt32();
+            public int Alignment => (int)((Flags & ValueWitnessFlags.AlignmentMask) + 1);
+            public bool IsNonPOD => Flags.HasFlag (ValueWitnessFlags.IsNonPOD);
+            public bool IsNonBitwiseTakable => Flags.HasFlag (ValueWitnessFlags.IsNonBitwiseTakable);
+            public bool HasExtraInhabitants => ExtraInhabitantCount != 0;
+        }
+
+        public static unsafe void* GetMetadata<T>(T type) where T: ISwiftObject
```

Comment:
Does this need to be generic?
```csharp
public static unsafe void* GetMetadata<T>(ISwiftObject obj) {
    return obj.Metadata;
}
```
While you don't use it here, we will need a more general version of this that takes the C# `Type` as an argument and returns the type metadata from that so that we can get the type metadata for the full scope of Swift projectable types.

Reply:
It is convenient for retrieval. Alternatively, it can be retrieved using reflection:
```csharp
        public static unsafe void* GetMetadata(Type t)
        {
            if (typeof(ISwiftObject).IsAssignableFrom(t))
            {
                var metadataProperty = t.GetProperty("Metadata", System.Reflection.BindingFlags.Static | System.Reflection.BindingFlags.Public);
                if (metadataProperty != null)
                {
                    var metadataValue = metadataProperty.GetValue(null);
                    if (metadataValue is IntPtr ptr)
                    {
                        return ptr.ToPointer();
                    }
                }
                throw new InvalidOperationException($"Type {t.Name} does not have a 'Metadata' property.");
            }
            throw new ArgumentException($"Type {t.Name} does not implement ISwiftObject.");
        }
```

I believe we should consider all use cases before making such changes, but I'm also open to refactoring it now. What do you prefer?

Reply:
This is going to take some thought, but I think the main entry point should be something like:
```
public static bool TryGetMetadata (Type t, out Metadata metadata)
{
    // one catch - t must not be an unbound generic.
    if (typeof(ISwiftObject).IsAssignableFrom(t)) {
        metadata = GetMetadataFromISwiftObject(t);
        return true;
    }
    // over time we add the cases we need for tuples, nint, nuint, nfloat, etc.
    // I'd like to see this ultimately be a ladder of `else if` constructs for each major type ordered by
    // most common and/or cheapest predicates. For example, identifying certain scalars should be
    // a matter of looking at `t.GetTypeCode()`
    return false;
}

static Metadata GetMetadataFromISwiftObject(Type t) // t is guaranteed to be ISwiftObject compatible 
{
    // your reflection code - no try/get pattern needed, this is a private API
}

Reply:
Yes, let's talk more about this. Created a tracking issue: https://github.com/dotnet/runtimelab/issues/2705